File: lib/__tests__/transactionMappers.test.ts
import { describe, expect, it } from 'vitest';
import {
  resolveBudgetIdFromRow,
  resolveBudgetNameForInsert,
  shouldSolidifyProjectedTransaction,
  toSupabaseTransaction,
} from '../hooks/transactionMappers';

const budgets = [
  { id: '11111111-1111-1111-1111-111111111111', name: 'Housing' },
  { id: 'budget:groceries', name: 'Groceries' },
  { id: 'custom-id', name: 'Emergency Fund' },
];

describe('shouldSolidifyProjectedTransaction', () => {
  // Use noon UTC so the test passes in every common timezone (UTC-12
  // through UTC+14). The old `00:01:00Z` was the previous local day in
  // any negative-offset timezone (e.g. America/Chicago), which made
  // the 'solidifies ... dated today' assertion fail there. The function
  // compares the tx date to the *local* date, so the test fixture has
  // to land on the same local day regardless of where it runs.
  const now = new Date('2026-03-20T12:00:00.000Z');

  it('solidifies projected transactions dated today', () => {
    expect(shouldSolidifyProjectedTransaction(true, '2026-03-20', now)).toBe(true);
  });

  it('solidifies projected transactions dated before today', () => {
    expect(shouldSolidifyProjectedTransaction(true, '2026-03-19', now)).toBe(true);
  });

  it('does not solidify projected transactions dated after today', () => {
    expect(shouldSolidifyProjectedTransaction(true, '2026-03-21', now)).toBe(false);
  });

  it('does not solidify non-projected transactions', () => {
    expect(shouldSolidifyProjectedTransaction(false, '2026-03-19', now)).toBe(false);
  });
});

describe('transactionMappers budget resolution', () => {
  it('returns null for rows with no Budget/budget field', () => {
    // The current schema (public.transactions) uses `Budget` (enum) /
    // `budget` (legacy alias). The `category_id` fallback was removed in
    // commit 63b244f because no live row in the current schema uses it.
    // resolveBudgetIdFromRow now returns null for rows that don't carry
    // a Budget/budget value, and the caller (toSupabaseTransaction)
    // is responsible for ensuring the transaction carries a budget_id
    // before it gets here.
    expect(resolveBudgetIdFromRow({})).toBeNull();
    expect(resolveBudgetIdFromRow({ category_id: 'abc-123' })).toBeNull();
  });

  it('maps legacy DB budget text values to stable system UUID ids', () => {
    expect(resolveBudgetIdFromRow({ budget: 'Housing' })).toBe('11111111-1111-1111-1111-111111111111');
  });

  it('prefers budget text over legacy category_id when both are present', () => {
    expect(resolveBudgetIdFromRow({ budget: 'Housing', category_id: 'stale-id' })).toBe(
      '11111111-1111-1111-1111-111111111111',
    );
  });

  it('falls back to normalized budget: prefixed ids for non-system categories', () => {
    expect(resolveBudgetIdFromRow({ budget: 'Emergency Fund' })).toBe('budget:emergency-fund');
  });

  it('resolves budget name from exact id for inserts', () => {
    expect(resolveBudgetNameForInsert('custom-id', budgets)).toBe('Emergency Fund');
  });

  it('resolves budget name from budget: prefixed id for inserts', () => {
    expect(resolveBudgetNameForInsert('budget:groceries', budgets)).toBe('Groceries');
  });

  it('resolves budget name when transaction carries category name as budget_id', () => {
    expect(resolveBudgetNameForInsert('Housing', budgets)).toBe('Housing');
  });

  it('throws when no matching budget can be found', () => {
    expect(() => resolveBudgetNameForInsert('missing', budgets)).toThrow(
      'Cannot map budget_id "missing" to a valid budget name for transactions.budget',
    );
  });

  it('maps a transaction budget_id to budget name for insert payloads', () => {
    const row = toSupabaseTransaction(
      {
        id: 'tx-1',
        user_id: 'user-1',
        vendor: 'Costco',
        amount: 12.34,
        date: '2026-03-09T12:00:00.000Z',
        budget_id: 'custom-id',
        recurrence: 'Monthly',
        label: 'Manual',
        is_projected: false,
        created_at: '2026-03-09T12:00:00.000Z',
      },
      budgets,
    );

    expect(row.budget).toBe('Emergency Fund');
    expect(row.recur).toBe('Monthly');
    expect(row.type).toBe('Manual');
  });
});

File: android-kotlin/app/src/test/java/com/covault/app/data/TransactionMappersTest.kt
package com.covault.app.data.remote

import com.covault.app.data.model.BudgetCategory
import com.covault.app.data.model.Recurrence
import com.covault.app.data.model.Transaction
import com.covault.app.data.model.TransactionLabel
import com.covault.app.data.model.TransactionSource
import com.covault.app.data.remote.dto.TransactionRow
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test
import java.time.LocalDate

class TransactionMappersTest {

    private val budgets = listOf(
        BudgetCategory("11111111-1111-1111-1111-111111111111", "Housing", 1500.0),
        BudgetCategory("22222222-2222-2222-2222-222222222222", "Groceries", 500.0),
        BudgetCategory("custom-emergency-fund", "Emergency Fund", 200.0),
    )

    private val today: LocalDate = LocalDate.of(2026, 3, 20)

    // --- shouldSolidifyProjectedTransaction --------------------------------

    @Test
    fun `solidifies projected transactions dated today`() {
        assertTrue(TransactionMappers.shouldSolidifyProjectedTransaction(
            isProjected = true, transactionDate = "2026-03-20", now = today,
        ))
    }

    @Test
    fun `solidifies projected transactions dated before today`() {
        assertTrue(TransactionMappers.shouldSolidifyProjectedTransaction(
            isProjected = true, transactionDate = "2026-03-19", now = today,
        ))
    }

    @Test
    fun `does not solidify projected transactions dated after today`() {
        assertEquals(false, TransactionMappers.shouldSolidifyProjectedTransaction(
            isProjected = true, transactionDate = "2026-03-21", now = today,
        ))
    }

    @Test
    fun `does not solidify non-projected transactions`() {
        assertEquals(false, TransactionMappers.shouldSolidifyProjectedTransaction(
            isProjected = false, transactionDate = "2026-03-19", now = today,
        ))
    }

    // --- resolveBudgetIdFromRow --------------------------------------------

    @Test
    fun `returns null for rows with no budget field`() {
        val row = TransactionRow(
            id = "tx-1", userId = "u-1", vendor = "Amazon", amount = 10.0,
            date = "2026-03-20", isProjected = false, budget = "",
            type = "Manual", recur = "One-time",
        )
        assertNull(TransactionMappers.resolveBudgetIdFromRow(row))
    }

    @Test
    fun `maps known DB budget text to system UUID`() {
        val row = TransactionRow(
            id = "tx-1", userId = "u-1", vendor = "Rent", amount = 1500.0,
            date = "2026-03-01", isProjected = false, budget = "Housing",
            type = "Automatic", recur = "Monthly",
        )
        assertEquals("11111111-1111-1111-1111-111111111111",
            TransactionMappers.resolveBudgetIdFromRow(row))
    }

    @Test
    fun `maps unknown DB budget text to budget prefix format`() {
        val row = TransactionRow(
            id = "tx-1", userId = "u-1", vendor = "Misc", amount = 5.0,
            date = "2026-03-01", isProjected = false, budget = "Pet Care",
            type = "Manual", recur = "One-time",
        )
        assertEquals("budget:pet-care", TransactionMappers.resolveBudgetIdFromRow(row))
    }

    // --- resolveBudgetNameForInsert ----------------------------------------

    @Test
    fun `resolves system UUID directly to budget name`() {
        assertEquals("Housing", TransactionMappers.resolveBudgetNameForInsert(
            "11111111-1111-1111-1111-111111111111", budgets,
        ))
    }

    @Test
    fun `resolves budget prefix id to name when name match found`() {
        assertEquals("Housing", TransactionMappers.resolveBudgetNameForInsert(
            "budget:housing", budgets,
        ))
    }

    @Test
    fun `throws on null budget id`() {
        try {
            TransactionMappers.resolveBudgetNameForInsert(null, budgets)
            fail("Expected IllegalArgumentException")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message!!.contains("valid budget_id"))
        }
    }

    @Test
    fun `throws when budget id cannot be mapped to a known name`() {
        try {
            TransactionMappers.resolveBudgetNameForInsert("totally-unknown", budgets)
            fail("Expected IllegalArgumentException")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message!!.contains("Cannot map budget_id"))
        }
    }

    // --- toSupabaseRow -----------------------------------------------------

    @Test
    fun `toSupabaseRow produces only columns the live schema supports`() {
        val tx = Transaction(
            id = "tx-1", userId = "u-1", vendor = "Amazon", amount = 49.99,
            date = "2026-03-20T12:00:00.000Z",
            budgetId = "22222222-2222-2222-2222-222222222222",
            recurrence = Recurrence.ONE_TIME,
            label = TransactionLabel.AUTOMATIC,
            isProjected = false, isIncome = false, caughtCleared = false,
            createdAt = "2026-03-20T12:00:00.000Z",
            source = TransactionSource.NOTIFICATION,
        )
        val row = TransactionMappers.toSupabaseRow(tx, budgets)
        assertEquals("tx-1", row["id"])
        assertEquals("u-1", row["user_id"])
        assertEquals("Amazon", row["vendor"])
        assertEquals(49.99, row["amount"])
        assertEquals("2026-03-20", row["date"])  // stripped to YYYY-MM-DD
        assertEquals(false, row["is_projected"])
        assertEquals("Groceries", row["budget"])
        assertEquals("Automatic", row["type"])
        assertEquals("One-time", row["recur"])
        assertEquals("notification", row["source"])
    }

    @Test
    fun `toSupabaseRow omits source when caller did not set it`() {
        val tx = Transaction(
            id = "tx-1", userId = "u-1", vendor = "Amazon", amount = 10.0,
            date = "2026-03-20", budgetId = "11111111-1111-1111-1111-111111111111",
            createdAt = "2026-03-20",
        )
        val row = TransactionMappers.toSupabaseRow(tx, budgets)
        assertTrue("source should be absent when null", !row.containsKey("source"))
    }

    @Test
    fun `toSupabaseRow throws when budget_id is null`() {
        val tx = Transaction(
            id = "tx-1", userId = "u-1", vendor = "Amazon", amount = 10.0,
            date = "2026-03-20", budgetId = null, createdAt = "2026-03-20",
        )
        try {
            TransactionMappers.toSupabaseRow(tx, budgets)
            fail("Expected IllegalArgumentException")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message!!.contains("valid budget_id"))
        }
    }

    // --- fromSupabaseRow ---------------------------------------------------

    @Test
    fun `fromSupabaseRow normalizes date to noon-UTC ISO`() {
        val row = TransactionRow(
            id = "tx-1", userId = "u-1", vendor = "Amazon", amount = 49.99,
            date = "2026-03-20", isProjected = false, budget = "Groceries",
            type = "Automatic", recur = "One-time",
        )
        val tx = TransactionMappers.fromSupabaseRow(row)
        assertEquals("2026-03-20T12:00:00.000Z", tx.date)
    }

    @Test
    fun `fromSupabaseRow solidifies past projected transactions`() {
        val row = TransactionRow(
            id = "tx-1", userId = "u-1", vendor = "Rent", amount = 1500.0,
            date = "2026-03-01", isProjected = true, budget = "Housing",
            type = "Automatic", recur = "Monthly",
            createdAt = "2026-02-25T00:00:00.000Z",
        )
        val tx = TransactionMappers.fromSupabaseRow(row)
        assertEquals(false, tx.isProjected)
    }

    @Test
    fun `fromSupabaseRow maps Automatic label and One-time recurrence`() {
        val row = TransactionRow(
            id = "tx-1", userId = "u-1", vendor = "Amazon", amount = 10.0,
            date = "2026-03-20", isProjected = false, budget = "Groceries",
            type = "Automatic", recur = "One-time", source = "notification",
            createdAt = "2026-03-20T00:00:00.000Z",
        )
        val tx = TransactionMappers.fromSupabaseRow(row)
        assertEquals(TransactionLabel.AUTOMATIC, tx.label)
        assertEquals(Recurrence.ONE_TIME, tx.recurrence)
        assertEquals(TransactionSource.NOTIFICATION, tx.source)
        assertNotNull(tx.budgetId)
    }
}
